External NIP-05 connection

martindsq opened the pull request:
#909 #910 #911

Implements option to connect your own nip05.

![Simulator Screenshot - iPhone 15 - 2024-03-22 at 16 25 01](https://github.com/planetary-social/nos/assets/1703242/73292e38-97ab-4f0a-bf91-31805fb44ac1)
![Simulator Screenshot - iPhone 15 - 2024-03-22 at 16 24 41](https://github.com/planetary-social/nos/assets/1703242/424b39ec-50ac-4169-80f6-0747aa4b0c7d)
![Simulator Screenshot - iPhone 15 - 2024-03-22 at 16 24 37](https://github.com/planetary-social/nos/assets/1703242/492f7c91-4530-4027-adf2-f3e89ab105dd)

mplorentz commented on Nos/Views/ProfileEdit/CreateUsernameWizard/ClaimYourUniqueIdentitySheet.swift:
```diff
@@ -15,7 +15,16 @@ struct ClaimYourUniqueIdentitySheet: View {
+                AlreadyHaveANIP05View(isPresented: $isPresented)
+            } label: {
+                SwiftUI.Text(.localizable.alreadyHaveANIP05)
+                    .font(.clarity(.medium, textStyle: .footnote))
```
I think we should bump this font up a size? The text size looks smaller in relation to the button above it than it does in Figma.

martindsq replied:
In Figma it's 14. Looking at https://developer.apple.com/design/human-interface-guidelines/typography: footnote is 13 and subheadline is 15. That's why I suggested the designs specify the dynamic type instead. I'll change it to subheadline because it's ambiguous.

mplorentz replied:
looks much better to my eyes now. Thanks!

mplorentz commented on Nos/Service/NamesAPI.swift:
Can you add a headerdoc comment to this class?

mplorentz commented on Nos/Service/NamesAPI.swift:
```diff
@@ -51,15 +51,33 @@ class NamesAPI {
+
+    /// Verifies that a given username_at_host NIP-05 can be connected to the keyPair
+    ///
+    /// - Parameter valueWhenNotFound: What to return if the server returns 404
```
What does it mean semantically when a 404 is returned? That the name is not already registered?

martindsq replied:
I have trouble with semantics here. When verifying nos.social usernames, we know that a username is free to take if the server returns a 404 or if it was already associated to the npub in the past. When verifying external nip-05, we are not looking for free to take usernames, we just want to check if it is already associated.

While they are semantically, the code its almost the same. I'll think a different approach.

martindsq replied:
I split verify into verify and checkAvailability. So, verify checks that a NIP-05 is recognized by the server while checkAvailability looks if a given username is free to take (server responded with 404) or was already recognizing that username for the npub is trying to check the availability.

mplorentz replied:
The new APIs are much easier for me to read, thanks!

mplorentz commented on Nos/Service/NamesAPI.swift:
```diff
@@ -51,15 +51,33 @@ class NamesAPI {
             let statusCode = response.statusCode
             if statusCode == 404 {
-                return true
+                return valueWhenNotFound
```
Maybe instead of having the caller indicate that they want this bool value to be flipped it would be less confusing if they just flipped it themselves at the call site?

martindsq replied:
But the caller won't know if it was a 404 or that the server didn't return an unexpected response.
A different approach would be to return a enum value so the called would know what was the server response.

mplorentz commented on Nos/Views/ProfileEdit/CreateUsernameWizard/ExternalNIP05/AlreadyHaveANIP05View.swift:
```diff
@@ -0,0 +1,199 @@
+                                }
+                            }
+                    }
+                    if linkFailed {
```
You could replace this `if ... else` block with `unableToLinkUsernameText().hidden(!linkFailed)`

martindsq replied:
`hidden()` doesn't accept a `bool`

mplorentz replied:
oh, haha. Well, it should 😅

mplorentz commented on Nos/Views/ProfileEdit/CreateUsernameWizard/ExternalNIP05/AlreadyHaveANIP05View.swift:
```diff
@@ -0,0 +1,199 @@
+        }
+    }
+
+    private func unableToLinkUsernameText() -> some View {
```
I'm not sure it matters but usually when we do subviews like this they are computed properties 🤷

mplorentz commented on Nos/Views/ProfileEdit/CreateUsernameWizard/ExternalNIP05/AlreadyHaveANIP05View.swift:
```diff
@@ -0,0 +1,199 @@
+            verified = try await namesAPI.verify(username: localPart, host: host, keyPair: keyPair)
+        } catch {
+            Log.error(error.localizedDescription)
+            verified = false
```
This line is redundant because of the above `defer`, right?

martindsq replied:
the `defer` modifies `isVerifying` not `verified`

mplorentz commented on Nos/Service/NamesAPI.swift:
```diff
@@ -51,15 +51,33 @@ class NamesAPI {
     }
 
+    /// Verifies that a given username is free to claim in nos.social
     func verify(username: String, keyPair: KeyPair) async throws -> Bool {
```
The comment explains it but I wish the function signature was more descriptive about what was being verified. Reading the call site makes it sound like we are verifying that username belongs to this keypair or something, but actually we are checking whether or not the given username is free (do we even need a keypair for this stage?). Maybe I am confusing things.

martindsq replied:
Yeah, you're right, let's think a different name

mplorentz commented on Nos/Service/NamesAPI.swift:
```diff
@@ -51,15 +51,33 @@ class NamesAPI {
+        host: URL,
+        keyPair: KeyPair,
+        valueWhenNotFound: Bool = false
+    ) async throws -> Bool {
```
I think this is pretty similar to `RelayService.verifyNIP05(...)`, maybe we can get rid of that one?

martindsq replied:
You're right, I missed that. Maybe we can use that instead?

martindsq replied:
I removed `RelayService.verifyNIP05`.

mplorentz reviewed (changes requested):
I had some trouble getting this one to work:
- When I tried to link my (valid) NIP-05 I saw an activity wheel for a second and then the button changed to "Next" but it is disabled.
- I tried going through the "Set up my @username" flow again to set up matt.nos.social and it said the name was already claimed. I think if someone already has a name reserved in our system then either flow (set up new name or existing name) should work for them to reclaim it.

martindsq commented:
> * When I tried to link my (valid) NIP-05 I saw an activity wheel for a second and then the button changed to "Next" but it is disabled.

Then it doesn't work, I'll need to keep working on it then.

> * I tried going through the "Set up my @username" flow again to set up matt.nos.social and it said the name was already claimed. I think if someone already has a name reserved in our system then either flow (set up new name or existing name) should work for them to reclaim it.

Yeah, it should be hable to reclaim it :(

martindsq commented:
@mplorentz 

> When I tried to link my (valid) NIP-05 I saw an activity wheel for a second and then the button changed to "Next" but it is disabled.

I fixed this

> I tried going through the "Set up my @username" flow again to set up matt.nos.social and it said the name was already claimed. I think if someone already has a name reserved in our system then either flow (set up new name or existing name) should work for them to reclaim it.
 
I tracked down this issue. It seems that our reserved usernames had to be manually added [using this script](https://github.com/planetary-social/nip05api/blob/main/scripts/add_name) in the server. When we delete the our nos.social usernames to test something, we have to re-add our reserver username again using the admin. You can ask @dcadenas to re-add yours again.

PS: I refactored `verify` so that the task using it can be cancelled while the user is still typing.

mplorentz reviewed (approved):
Sorry it took me so long to get this another review. It looks great now! I had one minor suggestion.
